command.go

```diff
@@ -71,58 +79,112 @@ func (c *Command) Run(ctx context.Context, stdout, stderr io.Writer) error {
 
 func RunAll(ctx context.Context, maxProcs int64, commands []*Command) error {
 	throttle := semaphore.NewWeighted(maxProcs)
-	errors := make(chan error, len(commands))
-	hasError := false
+	failures := make(chan *Failure, len(commands))
 	leftpad := 0
 	for _, command := range commands {
 		label := command.Label()
 		if len(label) > leftpad {
 			leftpad = len(label)
 		}
 	}
+	if leftpad < 3 {
+		leftpad = 3
+	}
 	leftpad++
 
 	for i, command := range commands {
-		throttle.Acquire(ctx, 1)
+		if err := throttle.Acquire(ctx, 1); err != nil {
+			return fmt.Errorf("acquire semaphore: %w", err)
+		}
+
 		go func(cmd *Command, i int) {
 			defer throttle.Release(1)
 
 			label := fmt.Sprintf("%*s |> ", leftpad, cmd.Label())
-
-			stdout, err := NewLogger(os.Stdout, label, i)
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				errors <- err
-				return
-			}
-			stderr, err := NewLogger(os.Stderr, label, i)
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				errors <- err
-				return
-			}
-
+			stdout := NewLogger(os.Stdout, label, i)
+			stderr := NewLogger(os.Stderr, label, i)
 			if err := cmd.Run(ctx, stdout, stderr); err != nil {
-				fmt.Fprintf(stderr, "%v", err)
-				errors <- err
+				writeNewlineAtEnd(stderr, err)
+				failures <- &Failure{
+					Command: cmd,
+					Error:   err,
+					Index:   i,
+				}
 			}
-			fmt.Fprintf(stderr, "\n")
 		}(command, i)
 	}
 	go func() {
 		// Ensure all processes have exited by acquiring maximum weight of semaphore...
-		throttle.Acquire(context.Background(), maxProcs)
+		_ = throttle.Acquire(context.Background(), maxProcs)
 		// ...before finally closing the channel, so the channel receiver doesn't terminate early.
-		close(errors)
+		close(failures)
 	}()
-	for err := range errors {
-		if err != nil {
-			hasError = true
+
+	// Buffer the report so that all threads can exit first.
+	var report bytes.Buffer
+	report.WriteRune('\n')
+	report.WriteRune('\n')
+
+	rw := &Logger{
+		w:      &report,
+		prefix: []byte(fmt.Sprintf("%*s |> ", leftpad, "***")),
+	}
+	color.New(color.Bold, color.BgBlack, color.FgHiBlack).Fprint(rw, "RUN REPORT")
```

Comment:
i have mixed feelings about the background color and text color combo here, feels like there isn't enough contrast?

Reply:
yeah the terminal background part is difficult. I think using bold font without the color alteration would be good enough for this? What do you think?

Reply:
Simplified to just using bold!